### PR TITLE
Add directory target support to copy sandbox

### LINKED ISSUE #7724: Copying Sandbox & Directory Targets
#7723 demonstrates that it's currently not possible to depend on a directory target whenever we sandbox a rule using copying.

### PR DESCRIPTION
**Pull request by emillon**
Fixes #7724

Signed-off-by: Etienne Millon <me@emillon.org>

### CODE REVIEW

**Review by rgrinberg — APPROVED** [review]

File: src/dune_engine/sandbox.ml [hunk 1]
```diff
@@ -41,6 +41,22 @@ type t =
 let dir t = t.dir
 let map_path t p = Path.Build.append t.dir p
 
+let rec copy_recursively (src_kind : Unix.file_kind) ~src ~dst =
```
Comment (rgrinberg) [inline src/dune_engine/sandbox.ml:44]: We have something similar to this in `Pkg_rules.Copy_tree`. Would be good to share implementations one day.

Comment (emillon) [reply src/dune_engine/sandbox.ml:44]: I tried to factor with `collect_dir_recursively` too but it didn't go well.

File: test/blackbox-tests/test-cases/directory-targets/copy-sandboxing.t [hunk 2]
```diff
@@ -14,10 +14,11 @@ mode:
   >  (target foo)
   >  (deps (sandbox always) output/)
   >  (action
-  >   (run touch foo)))
+  >   (progn
+  >    (run ls output)
```
[the remark this answers is not in the document]
Comment (emillon) [reply test/blackbox-tests/test-cases/directory-targets/copy-sandboxing.t:18]: should I add that separately as `dune_cmd ls`?

[the remark this answers is not in the document]
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/directory-targets/copy-sandboxing.t:18]: If you'd like. Though we use `ls output | sort` or `find output | sort` as an idiom pretty often.

**Comment by rgrinberg** [comment]
@pmwhite do you want to pull this fix?

**Comment by pmwhite** [comment]
This has been pulled.

**Pull request merged**